new/language/implementation/javascript/javascript.go

```diff
@@ -0,0 +1,240 @@
+package javascript
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/smacker/go-tree-sitter/javascript"
+	"github.com/ssoroka/slice"
+	"golang.org/x/exp/slices"
+
+	"github.com/bearer/curio/new/language/implementation"
+	"github.com/bearer/curio/new/language/tree"
+	"github.com/bearer/curio/pkg/util/regex"
+
+	patternquerytypes "github.com/bearer/curio/new/language/patternquery/types"
+	sitter "github.com/smacker/go-tree-sitter"
+)
+
+var (
+	variableLookupParents = []string{"pair", "arguments", "binary_expression"}
+
+	anonymousPatternNodeParentTypes = []string{}
+	patternMatchNodeContainerTypes  = []string{}
+
+	// $<name:type> or $<name:type1|type2> or $<name>
+	patternQueryVariableRegex = regexp.MustCompile(`\$<(?P<name>[^>:!]+)(?::(?P<types>[^>]+))?>`)
+	allowedPatternQueryTypes  = []string{"identifier", "property_identifier", "_", "member_expression"}
+
+	matchNodeRegex = regexp.MustCompile(`\$<!>`)
+
+	ellipsisRegex = regexp.MustCompile(`\$<\.\.\.>`)
+)
+
+type javascriptImplementation struct{}
+
+func Get() implementation.Implementation {
+	return &javascriptImplementation{}
+}
+
+func (implementation *javascriptImplementation) SitterLanguage() *sitter.Language {
+	return javascript.GetLanguage()
+}
+
+func (implementation *javascriptImplementation) AnalyzeFlow(rootNode *tree.Node) error {
+	scope := make(map[string]*tree.Node)
+
+	return rootNode.Walk(func(node *tree.Node, visitChildren func() error) error {
+		switch node.Type() {
+		case "function":
+			scope = make(map[string]*tree.Node)
+		// () => {}
+		case "arrow_function":
+			scope = make(map[string]*tree.Node)
+		// function getName() {}
+		case "method_definition":
+			scope = make(map[string]*tree.Node)
+		// user = ...
+		case "assignment_expression":
+			left := node.ChildByFieldName("left")
+			right := node.ChildByFieldName("right")
+
+			if left.Type() == "identifier" {
+				err := visitChildren()
+
+				scope[left.Content()] = node
+				node.UnifyWith(right)
+
+				return err
+			}
+		// const user = ...
+		// var user = ...
+		// let user = ...
+		case "variable_declarator":
+			name := node.ChildByFieldName("name")
+			value := node.ChildByFieldName("value")
+
+			if name.Type() == "identifier" {
+				err := visitChildren()
+
+				scope[name.Content()] = node
+				node.UnifyWith(value)
+
+				return err
+			}
+		case "identifier":
+			parent := node.Parent()
+			if parent != nil && slice.Contains(variableLookupParents, parent.Type()) {
+				scopedNode := scope[node.Content()]
+				if scopedNode != nil {
+					node.UnifyWith(scopedNode)
+				}
+			}
+		case "property_identifier":
```

Comment:
I'm not sure we want this case. I think property identifier is just a key name isn't it? ie. not a variable reference.